### PR TITLE
fix: rate limit during json-ld context loading

### PR DESCRIPTION
**Pull request by DominikPinsel**
Fix rate limit during json-ld context loading.

Changes
- make use of DidDocumentResolver from the ssi library 0.0.17
  - this version supports context caching
  - had to change some things, so that the code still compiles
- added logging for all incoming http calls
  - it was not possible to see what the MIW was doing on int
- repaired tests
  - added a valid did to a test
  - added a default vc expiry date to test setup

### CODE REVIEW

**Review by florianrusch-zf — CHANGES_REQUESTED** [review]

**Comment by florianrusch-zf** [comment]
Btw. why is there no GitHub Issue for this?

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/service/IssuersCredentialService.java [hunk 1]
```diff
@@ -430,17 +430,9 @@ public Map<String, Object> credentialsValidation(Map<String, Object> data, boole
...
-            proofValidation = LinkedDataProofValidation.newInstance(SignatureType.ED21559, didResolver);
-        } else if (SignatureType.JWS.toString().equals(proofTye)) {
-            proofValidation = LinkedDataProofValidation.newInstance(SignatureType.JWS, didResolver);
-        } else {
-            throw new BadDataException(String.format("Invalid proof type: %s", proofTye));
-        }
+        LinkedDataProofValidation proofValidation = LinkedDataProofValidation.newInstance(didResolver);
 
-        boolean valid = proofValidation.verifyProof(verifiableCredential);
+        boolean valid = proofValidation.verifiy(verifiableCredential);
```
[the remark this answers is not in the document]
Comment (borisrizov-zf) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/service/IssuersCredentialService.java:435]: This was fixed in an older version, please recheck @DominikPinsel

File: src/test/java/org/eclipse/tractusx/managedidentitywallets/vc/HoldersCredentialTest.java [hunk 2]
```diff
@@ -290,9 +289,9 @@ void validateExpiredCredentialsWithExpiryCheckTrue() throws com.fasterxml.jackso
             //mock setup
             LinkedDataProofValidation mock = Mockito.mock(LinkedDataProofValidation.class);
             utils.when(() -> {
-                LinkedDataProofValidation.newInstance(Mockito.any(SignatureType.class), Mockito.any(DidResolver.class));
+                LinkedDataProofValidation.newInstance(Mockito.any(DidResolver.class));
             }).thenReturn(mock);
-            Mockito.when(mock.verifyProof(Mockito.any(VerifiableCredential.class))).thenReturn(true);
+            Mockito.when(mock.verifiy(Mockito.any(VerifiableCredential.class))).thenReturn(true);
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/test/java/org/eclipse/tractusx/managedidentitywallets/vc/HoldersCredentialTest.java:294]: This type is part of the library. I could fix it, but then the code won't compile anymore :)

File: src/test/java/org/eclipse/tractusx/managedidentitywallets/vc/HoldersCredentialTest.java [hunk 3]
```diff
@@ -264,9 +263,9 @@ void validateCredentialsWithExpiryCheckFalse() throws com.fasterxml.jackson.core
...
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/test/java/org/eclipse/tractusx/managedidentitywallets/vc/HoldersCredentialTest.java:268]: This type is part of the library. I could fix it, but then the code won't compile anymore :)

File: src/test/java/org/eclipse/tractusx/managedidentitywallets/vc/HoldersCredentialTest.java [hunk 4]
```diff
@@ -237,9 +236,9 @@ void validateCredentialsWithExpiryCheckTrue() throws com.fasterxml.jackson.core.
...
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/test/java/org/eclipse/tractusx/managedidentitywallets/vc/HoldersCredentialTest.java:241]: This type is part of the library. I could fix it, but then the code won't compile anymore :)

File: src/test/java/org/eclipse/tractusx/managedidentitywallets/vc/HoldersCredentialTest.java [hunk 5]
```diff
@@ -214,9 +213,9 @@ void validateCredentialsWithInvalidVC() throws com.fasterxml.jackson.core.JsonPr
             //mock setup
             LinkedDataProofValidation mock = Mockito.mock(LinkedDataProofValidation.class);
             utils.when(() -> {
-                LinkedDataProofValidation.newInstance(Mockito.any(SignatureType.class), Mockito.any(DidResolver.class));
+                LinkedDataProofValidation.newInstance(Mockito.any(DidResolver.class));
             }).thenReturn(mock);
-            Mockito.when(mock.verifyProof(Mockito.any(VerifiableCredential.class))).thenReturn(false);
+            Mockito.when(mock.verifiy(Mockito.any(VerifiableCredential.class))).thenReturn(false);
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/test/java/org/eclipse/tractusx/managedidentitywallets/vc/HoldersCredentialTest.java:218]: This type is part of the library. I could fix it, but then the code won't compile anymore :)

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/service/PresentationService.java [hunk 6]
```diff
@@ -262,25 +265,10 @@ private boolean validateAudience(String audience, SignedJWT signedJWT) {
...
-            linkedDataProofValidation = LinkedDataProofValidation.newInstance(
-                    SignatureType.JWS,
-                    didResolver
-            );
-        } else {
-            throw new UnsupportedSignatureTypeException(proofType);
-        }
+        final DidResolver resolver = didDocumentResolverService.getCompositeDidResolver();
+        final LinkedDataProofValidation linkedDataProofValidation = LinkedDataProofValidation.newInstance(resolver);
+        final boolean isValid = linkedDataProofValidation.verifiy(credential);
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/service/PresentationService.java:270]: This type is part of the library. I could fix it, but then the code won't compile anymore :)

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/ManagedIdentityWalletsApplication.java [hunk 7]
```diff
@@ -21,26 +21,73 @@
...
  */
 @SpringBootApplication
 @ConfigurationPropertiesScan
 @EnableTransactionManagement
+@Slf4j
 public class ManagedIdentityWalletsApplication {
 
+    private static final URI URI_BUSINESS_VC_V1 = URI.create("https://catenax-ng.github.io/product-core-schemas/businessPartnerData.json");
+    private static final String RESOURCE_BUSINESS_VC_V1 = "/jsonld/cx_business_vc_v1.json";
+    private static final URI URI_SUMMARY_VC_V1 = URI.create("https://catenax-ng.github.io/product-core-schemas/SummaryVC.json");
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/ManagedIdentityWalletsApplication.java:50]: This is true, but as it is now the Catena-X schemas are stored there. We already addressed that they should be resolved using a real Catena-X domain, but this is not the case today.

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java [hunk 8]
```diff
@@ -174,6 +177,7 @@ public ResponseEntity<DidDocument> getDidDocument(@Parameter(description = "Did
     @Operation(description = "Resolve the DID document for a given BPN", summary = "Resolve DID Document")
     @GetMapping(path = RestURI.DID_RESOLVE, produces = MediaType.APPLICATION_JSON_VALUE)
     public ResponseEntity<DidDocument> getDidResolve(@Parameter(description = "BPN",examples = {@ExampleObject(name = "bpn", value = "BPNL000000000000", description = "bpn")}) @PathVariable(name = "bpn") String bpn) {
+        log.info("Get DID document for bpn: {}", bpn);
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java]: Yea, you are right. I used the info level to test something in Hotel Budapest and then forgot to adjust it before creating the PR. Thank you for the review :)

- Updated the log messages of the DidDocumentController
- Set all the log levels to debug for receiving http requests

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/service/IssuersCredentialService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/service/IssuersCredentialService.java:435]: Rechecked. It only compiles if written like this :)

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/service/PresentationService.java [hunk 9]
```diff
@@ -240,7 +241,9 @@ private boolean validateJWTExpiryDate(SignedJWT signedJWT) {
             jwtValidator.validateDate(signedJWT);
             return true;
         } catch (Exception e) {
-            log.error("Can not expiry date ", e);
+            if ((!(e instanceof JwtExpiredException))) {
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/service/PresentationService.java:244]: Yes there is :) Removed one of the outer brackets.

**Comment by DominikPinsel** [comment]
> Btw. why is there no GitHub Issue for this?

Created Issue #103

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java [hunk 10]
```diff
@@ -109,6 +111,7 @@ public class DidDocumentController {
     @Operation(description = "Resolve the DID document for a given DID or BPN", summary = "Resolve DID Document")
     @GetMapping(path = RestURI.DID_DOCUMENTS, produces = MediaType.APPLICATION_JSON_VALUE)
     public ResponseEntity<DidDocument> getDidDocument(@Parameter(description = "Did or BPN",examples = {@ExampleObject(name = "bpn", value = "BPNL000000000000", description = "bpn"), @ExampleObject(description = "did", name = "did", value = "did:web:localhost:BPNL000000000000")}) @PathVariable(name = "identifier") String identifier) {
+        log.debug("Received Request to get DID document for identifier: {}", identifier);
```
Comment (florianrusch-zf) [inline src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java:114]: Let's keep it consistent
```suggestion
        log.debug("Received request to get DID document for identifier: {}", identifier);
```

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java [hunk 11]
```diff
@@ -174,6 +177,7 @@ public ResponseEntity<DidDocument> getDidDocument(@Parameter(description = "Did
     @Operation(description = "Resolve the DID document for a given BPN", summary = "Resolve DID Document")
     @GetMapping(path = RestURI.DID_RESOLVE, produces = MediaType.APPLICATION_JSON_VALUE)
     public ResponseEntity<DidDocument> getDidResolve(@Parameter(description = "BPN",examples = {@ExampleObject(name = "bpn", value = "BPNL000000000000", description = "bpn")}) @PathVariable(name = "bpn") String bpn) {
+        log.debug("Received Request to get DID document for identifier: {}", bpn);
```
Comment (florianrusch-zf) [inline src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java:180]:
```suggestion
        log.debug("Received request to get DID document for identifier: {}", bpn);
```

**Review by florianrusch-zf — CHANGES_REQUESTED** [review]

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/service/IssuersCredentialService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (florianrusch-zf) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/service/IssuersCredentialService.java:435]: Please talk to each other. Within the [PR](https://github.com/eclipse-tractusx/SSI-agent-lib/pull/15) from @borisrizov-zf it got renamed.

[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/service/IssuersCredentialService.java:435]: Here is the code from the main branch in catena-x ng.
https://github.com/eclipse-tractusx/SSI-agent-lib/blob/main/src/main/java/org/eclipse/tractusx/ssi/lib/proof/LinkedDataProofValidation.java

The Tractus-X ssi repository does not release artifacts yet and seems to be in another state.

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/ManagedIdentityWalletsApplication.java [hunk 12]
```diff
@@ -21,26 +21,73 @@
...
      * @param args the input arguments
      */
     public static void main(String[] args) {
+        initializeOfflineResources();
+
         SpringApplication.run(ManagedIdentityWalletsApplication.class, args);
     }
 
+    static void initializeOfflineResources() {
+        RemoteDocumentLoader.getInstance().setEnableLocalCache(true);
```
Comment (borisrizov-zf) [inline src/main/java/org/eclipse/tractusx/managedidentitywallets/ManagedIdentityWalletsApplication.java:69]: Is there actually a mechanism to force update the resources without a PR?

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java [hunk 13]
```diff
@@ -174,6 +177,7 @@ public ResponseEntity<DidDocument> getDidDocument(@Parameter(description = "Did
     @Operation(description = "Resolve the DID document for a given BPN", summary = "Resolve DID Document")
     @GetMapping(path = RestURI.DID_RESOLVE, produces = MediaType.APPLICATION_JSON_VALUE)
     public ResponseEntity<DidDocument> getDidResolve(@Parameter(description = "BPN",examples = {@ExampleObject(name = "bpn", value = "BPNL000000000000", description = "bpn")}) @PathVariable(name = "bpn") String bpn) {
+        log.debug("Received request to get DID document for identifier: {}", bpn);
```
Comment (borisrizov-zf) [inline src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java:180]:
```suggestion
        log.debug("Received request to get DID document for BPN: {}", bpn);
```

**Review by borisrizov-zf — COMMENTED** [review]
lgtm so far, I'll test locally as well and give a final approval.

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/service/IssuersCredentialService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (borisrizov-zf) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/service/IssuersCredentialService.java:435]: @florianrusch-zf yes, the tractus-x repo is up-to-date, but not yet ready to be used as it has no artefact. we'll have to use the catenax-ng repo for a couple PRs before we can finalise and move.

**Comment by borisrizov-zf** [comment]
This must be merged before #87 as it fixes some issues causing the pipeline to break.

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java]: As the MIW Is able to resolve DID documents by BPN and DID (in the API called identifier) I assume this is really correct.

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/ManagedIdentityWalletsApplication.java (hunk 12 above)
Comment (DominikPinsel) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/ManagedIdentityWalletsApplication.java:69]: The schemas, we put in the local cache, can only be updated by a PR. I would assume this is ok, as the resources itself should not change once released. An update should happen by publishing a new version.

File: src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java (hunk 13 above)
Comment (DominikPinsel) [reply src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/DidDocumentController.java:180]: The MIW is able to resovle a DID Document By BPN and DID (in the API called identifier). So naming it here identifier is somewhat correct, although confusing.

**Comment by borisrizov-zf** [comment]
I've given this another thought. I think it won't be wise to implement another level of caching in the MIW. The SSI-Agent-lib already includes a 1 day cache mechanism for the context files, which should be enough. Also using a hard local resource seems unflexible if any changes to the schema happen. Let's shelve this for now.

**Comment by borisrizov-zf** [comment]
> LGFM, but you (@borisrizov-zf @DominikPinsel) have to decided whether this should be merged or not.

I think we should abandon, because having local copies of the schemas will come back to bite us in the backside, imo.

**Comment by DominikPinsel** [comment]
> > LGFM, but you (@borisrizov-zf @DominikPinsel) have to decided whether this should be merged or not.
> 
> I think we should abandon, because having local copies of the schemas will come back to bite us in the backside, imo.

So close PR?

**Comment by borisrizov-zf** [comment]
> > > LGFM, but you (@borisrizov-zf @DominikPinsel) have to decided whether this should be merged or not.
> > 
> > 
> > I think we should abandon, because having local copies of the schemas will come back to bite us in the backside, imo.
> 
> So close PR?

Yeah, close it.

**Comment by borisuu** [comment]
> Update: Reopen and will just remove the offline resources

Ok, ping me when it's done for review.

**Comment by DominikPinsel** [comment]
Removed offline loading, squashed changes

@florianrusch-zf @borisrizov-zf FYI

**Review by borisrizov-zf — APPROVED** [review]

**Comment by borisrizov-zf** [comment]
@florianrusch-zf this is ready for merging. Thanks.

**Comment by arunsworld** [comment]
> I've given this another thought. I think it won't be wise to implement another level of caching in the MIW. The SSI-Agent-lib already includes a 1 day cache mechanism for the context files, which should be enough. Also using a hard local resource seems unflexible if any changes to the schema happen. Let's shelve this for now.

Nothing in the documentation to indicate that egress to a certain set of URLs is a pre-requisite to run MIW. It's a normal thing that firewalls block egress. You're worried about schema changes not getting picked up? Are we really saying schemas will change without version changing? That in itself is unsettling (left-pad that broke the internet - at least that was build time; this is runtime). MIW is core to all EDC transactions; this looks like an Achilles heel / single point of failure. I hope you reconsider putting this back in or at least having an option to do so. How about starting with a local copy and refreshing on startup and then daily if possible. But no matter what you have something to work with and it's not going to stop the world.

**Comment by borisrizov-zf** [comment]
> > I've given this another thought. I think it won't be wise to implement another level of caching in the MIW. The SSI-Agent-lib already includes a 1 day cache mechanism for the context files, which should be enough. Also using a hard local resource seems unflexible if any changes to the schema happen. Let's shelve this for now.
> 
> Nothing in the documentation to indicate that egress to a certain set of URLs is a pre-requisite to run MIW. It's a normal thing that firewalls block egress. You're worried about schema changes not getting picked up? Are we really saying schemas will change without version changing? That in itself is unsettling (left-pad that broke the internet - at least that was build time; this is runtime). MIW is core to all EDC transactions; this looks like an Achilles heel / single point of failure. I hope you reconsider putting this back in or at least having an option to do so. How about starting with a local copy and refreshing on startup and then daily if possible. But no matter what you have something to work with and it's not going to stop the world.

The underlying SSI library already adheres to your proposed workflow. We get the resources and cache them for one day. I think it would be reasonable to add static resources when the schemas are in their final state. We're still experiencing changes in those, so makes no sense to update the repo every time there's a change. We might open a task to remind ourselves to add static resources whenever a final version is available.

**Pull request merged**